remote_test.go

```diff
@@ -189,6 +190,73 @@ func (s *RemoteSuite) TestFetchWithDepth(c *C) {
 	c.Assert(r.s.(*memory.Storage).Objects, HasLen, 18)
 }
 
+func (s *RemoteSuite) TestFetchWithHashes(c *C) {
+	r := NewRemote(memory.NewStorage(), &config.RemoteConfig{
+		URLs: []string{s.GetBasicLocalRepositoryURL()},
+	})
+
+	s.testFetch(c, r, &FetchOptions{
+		Depth: 1,
+		Hashes: []plumbing.Hash{
+			plumbing.NewHash("6ecf0ef2c2dffb796033e5a02219af86ec6584e5"),
+		},
+	}, nil)
+
+	commits := r.s.(*memory.Storage).Commits
+	c.Assert(len(commits), Equals, 1)
+}
+
+func (s *RemoteSuite) TestFetchWithHashesDepthOfTwo(c *C) {
+	r := NewRemote(memory.NewStorage(), &config.RemoteConfig{
+		URLs: []string{s.GetBasicLocalRepositoryURL()},
+	})
+
+	s.testFetch(c, r, &FetchOptions{
+		Depth: 2,
+		Hashes: []plumbing.Hash{
+			plumbing.NewHash("6ecf0ef2c2dffb796033e5a02219af86ec6584e5"),
+		},
+	}, nil)
+
+	commits := r.s.(*memory.Storage).Commits
+	c.Assert(len(commits), Equals, 2)
+}
+
+func (s *RemoteSuite) TestFetchWithHashesAndReferences(c *C) {
+	r := NewRemote(memory.NewStorage(), &config.RemoteConfig{
+		URLs: []string{s.GetBasicLocalRepositoryURL()},
+	})
+
+	s.testFetch(c, r, &FetchOptions{
+		Depth: 1,
+		Hashes: []plumbing.Hash{
+			plumbing.NewHash("6ecf0ef2c2dffb796033e5a02219af86ec6584e5"),
+		},
+		RefSpecs: []config.RefSpec{
+			config.RefSpec("+refs/heads/branch:refs/remotes/origin/branch"),
+		},
+	}, []*plumbing.Reference{
+		plumbing.NewReferenceFromStrings("refs/remotes/origin/branch", "e8d3ffab552895c19b9fcf7aa264d277cde33881"),
+	})
+	commits := r.s.(*memory.Storage).Commits
+	c.Assert(len(commits), Equals, 2)
```

Comment:
A comment explaining the scenario here would be cool. I'm assuming that we're getting 1 commit from the shallow ref and the other one is from the bare oid. Is the bare one the parent of the branch head or is it further back in history? It would be cool to have the 2 commits be non-adjacent since that's what we're going for.

Reply:
The SHAs in this test come from two different branches:

- https://github.com/git-fixtures/basic/commits/master
- https://github.com/git-fixtures/basic/commits/branch

Coincidentally, they both have the same parent, but this fetch operation does not result in that parent being pulled down.